portable-atomic-util/src/arc.rs

```diff
@@ -0,0 +1,859 @@
+
+    unsafe fn from_inner(ptr: NonNull<Shared<T>>) -> Self {
+        Self { shared: ptr }
+    }
+
+    /// Try to upgrade this `Weak` pointer to a strong pointer.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use portable_atomic_util::Arc;
+    ///
+    /// let five = Arc::new(5);
+    ///
+    /// let weak = Arc::downgrade(&five);
+    /// assert!(weak.upgrade().is_some());
+    /// ```
+    #[must_use]
+    pub fn upgrade(&self) -> Option<Arc<T>> {
+        let header = self.inner()?;
+        let mut strong = header.strong.load(Relaxed);
+
+        loop {
+            // If the strong count is zero, the allocation has been dropped.
+            if strong == 0 {
+                return None;
+            }
+
+            // If the strong count is greater than the maximum, abort.
+            if strong > MAX_REFCOUNT {
+                abort();
+            }
+
+            // Try to increment the strong count.
+            match header.strong.compare_exchange_weak(strong, strong + 1, Acquire, Relaxed) {
+                Ok(_) => {
+                    // SAFETY: Now that the strong counter is incremented, the allocation for an Arc is valid.
+                    return unsafe { Some(Arc::from_inner(self.shared)) };
+                }
+                Err(x) => strong = x,
+            }
+        }
+    }
+
+    /// Get the number of strong pointers to this allocation.
+    #[must_use]
+    pub fn strong_count(&self) -> usize {
+        self.inner().map_or(0, |header| header.strong.load(Acquire))
+    }
+
+    /// Get the number of weak pointers to this allocation.
+    #[must_use]
+    pub fn weak_count(&self) -> usize {
+        self.inner().map_or(0, |header| {
+            let weak = header.weak.load(Acquire);
+            let strong = header.strong.load(Acquire);
+
+            if strong == 0 {
+                0
+            } else {
+                weak - 1
+            }
+        })
+    }
+}
+
+impl<T> Weak<T> {
+    /// Create a new dangling `Weak` pointer.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use portable_atomic_util::Weak;
+    ///
+    /// let weak = Weak::<i32>::new();
+    /// ```
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            // SAFETY: usize::MAX != 0, so this is always valid
+            shared: unsafe { NonNull::new_unchecked(strict::invalid(core::usize::MAX)) },
+        }
+    }
+}
+
+impl<T> Default for Weak<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T: ?Sized> Drop for Weak<T> {
+    fn drop(&mut self) {
+        // Try to get access to the inner shared pointer.
+        if let Some(inner) = self.inner() {
+            if inner.weak.fetch_sub(1, Release) == 1 {
+                acquire!(inner.weak);
+
+                // Deallocate the memory.
+                //
+                // SAFETY: We know that the weak count is 0, so we can deallocate.
+                drop(unsafe { Box::<Shared<T>>::from_raw(self.shared.as_ptr()) });
```

Comment:
I think this might be a soundness hole, on second glance. Let me try changing it to deallocate without dropping `T`.